**If you lost points on the last checkpoint you can get them back by responding to TA/IA feedback**  

Update/change the relevant sections where you lost those points, make sure you respond on GitHub Issues to your TA/IA to call their attention to the changes you made here.

Please update your Timeline... no battle plan survives contact with the enemy, so make sure we understand how your plans have changed.

# COGS 108 - EDA Checkpoint

# Names

- Andy Huang
- Audrey Kim
- Katie Moc
- Khang Quach
- Elton Villalta

# Research Question

-  Include a specific, clear data science question.
-  Make sure what you're measuring (variables) to answer the question is clear

What is your research question? Include the specific question you're setting out to answer. This question should be specific, answerable with data, and clear. A general question with specific subquestions is permitted. (1-2 sentences)

____

Given a current year’s top hit songs, is there a correlation between genre, popularity, and lyrical density to any established swing states’ final vote?


## Background and Prior Work


- Include a general introduction to your topic
- Include explanation of what work has been done previously
- Include citations or links to previous work

This section will present the background and context of your topic and question in a few paragraphs. Include a general introduction to your topic and then describe what information you currently know about the topic after doing your initial research. Include references to other projects who have asked similar questions or approached similar problems. Explain what others have learned in their projects.

Find some relevant prior work, and reference those sources, summarizing what each did and what they learned. Even if you think you have a totally novel question, find the most similar prior work that you can and discuss how it relates to your project.

References can be research publications, but they need not be. Blogs, GitHub repositories, company websites, etc., are all viable references if they are relevant to your project. It must be clear which information comes from which references. (2-3 paragraphs, including at least 2 references)

 **Use inline citation through HTML footnotes to specify which references support which statements** 

For example: After government genocide in the 20th century, real birds were replaced with surveillance drones designed to look just like birds.<a name="cite_ref-1"></a>[<sup>1</sup>](#cite_note-1) Use a minimum of 2 or 3 citations, but we prefer more.<a name="cite_ref-2"></a>[<sup>2</sup>](#cite_note-2) You need enough to fully explain and back up important facts. 

Note that if you click a footnote number in the paragraph above it will transport you to the proper entry in the footnotes list below.  And if you click the ^ in the footnote entry, it will return you to the place in the main text where the footnote is made.

To understand the HTML here, `<a name="#..."> </a>` is a tag that allows you produce a named reference for a given location.  Markdown has the construciton `[text with hyperlink](#named reference)` that will produce a clickable link that transports you the named reference.

1. <a name="cite_note-1"></a> [^](#cite_ref-1) Lorenz, T. (9 Dec 2021) Birds Aren’t Real, or Are They? Inside a Gen Z Conspiracy Theory. *The New York Times*. https://www.nytimes.com/2021/12/09/technology/birds-arent-real-gen-z-misinformation.html 
2. <a name="cite_note-2"></a> [^](#cite_ref-2) Also refs should be important to the background, not some randomly chosen vaguely related stuff. Include a web link if possible in refs as above.

____

Throughout history, music has played an important role in the creative expression of disadvantaged communities. We want to gain a greater insight of this relationship using exploratory data analysis. After doing some research, what we currently know is that groups fighting for liberation often turned to the rap and R&B genres and wrote songs about their struggles, especially throughout the civil rights movement. In more recent decades, we have seen increased political expression in other genres, such as pop and singer-songwrier. We want to find out if these genres and other music variables have a correlation with political affiliation over time by state.

Past research has shown that people's music preferences can help denote their political affiliation. Specifically, republicans typically prefer country music and democrats typically prefer pop, rap, hip-hop, classic rock, and/or alternative music. Additionally, democrats are less likely to prefer country music.<a name="cite_ref-1"></a>[<sup>1</sup>](#cite_note-3) Further studies delve into the relationship between music preferences, partisanship, and political attitudes. Historically, there has been a connection between music and political affiliation, where music often has an influence and reflected political events. Using an original survey for newer findings, the study again found that Republicans are more likely to prefer Country music, while Democrats tend to more Classic Rock and Alternative genres. These preferences were found to influence the person's political attitudes, where Republicans preferring New Country, shows more favored opinions of Republican Congressmen and the Supreme Court. These findings show that political identities can be strengthened with music preferences through Affective polarization.<a name="cite_ref-1"></a>[<sup>2</sup>](#cite_note-3) This is relevant to our project because this research showcases a clear correlation between music genre and political party. We want to explore this further by researching additional variables and focus on the trends among various US states.

1. <a name="cite_note-3"></a> [^](#cite_ref-3) Dolan, Eric W. “Music Preferences Serve as Markers of Political Affiliation.” PsyPost, 2 Mar. 2024, https://www.psypost.org/music-preferences-serve-as-markers-of-political-affiliation/
2. <a name="cite_note-4"></a> [^](#cite_ref-4) Mack, Brianna N. Martin, Teresa R. 26 December, 2023. "Party Rocking: Exploring the relationship between music preference, partisanship, and political attitudes" Ohio Wesleyan University. https://www.sciencedirect.com/science/article/pii/S0304422X23001018

# Hypothesis



- Include your team's hypothesis
- Ensure that this hypothesis is clear to readers
- Explain why you think this will be the outcome (what was your thinking?)

What is your main hypothesis/predictions about what the answer to your question is? Briefly explain your thinking. (2-3 sentences)

_____

We hypothesize that given a year’s top hit songs, we can predict a given swing states’ final vote due to high correlations between genres, popularity and lyrical density.

During an election year, if the most popular genres are classic rock and alternative with high lyrical density we can predict that a swing states’ final vote to be Democratic, while years where the most popular genres are country and classical with less lyrical density tend to lean more Republican. As such, given the popularity of a song with its genre and lyrical density, songs of higher popularity are weighted more as they tend to be a better predictor of the direction of the swing state’s final vote.

# Data

## Data overview

For each dataset include the following information
- Dataset #1
  - Dataset Name: Spotify-Data 1921-2020 (data.csv)
  - Link to the dataset: https://www.kaggle.com/datasets/ektanegi/spotifydata-19212020
  - Number of observations: 169909
  - Number of variables: 19
- Dataset #2 (if you have more than one!)
  - Dataset Name: US Elections Dataset (1976-2020-president.csv)
  - Link to the dataset: https://www.kaggle.com/datasets/tunguz/us-elections-dataset/data 
  - Number of observations: 4287
  - Number of variables: 15

Now write 2 - 5 sentences describing each dataset here. Include a short description of the important variables in the dataset; what the metrics and datatypes are, what concepts they may be proxies for. Include information about how you would need to wrangle/clean/preprocess the dataset

If you plan to use multiple datasets, add a few sentences about how you plan to combine these datasets.

_______

For both of our datasets we will first clean them both by getting rid of all null values, this will ensure there are no holes in our data that could cause discrepancies in our analysis. After doing that we will then confirm that all the data are formatted in a similar fashion, ie: years are all within XXXX form, names of States are formatted the same without duplicates (CA == California), metrics are all the same form (everything are floats, strings, ints) etc. After we do general data cleaning we will then only use the years from 2011 onwards due to the fact that Spotify was released within the US in 2011 thus all years before 2011 will not be considered, we will then finally pull out the corresponding columns that we are interested in.

For the dataset that contains the Top Hits from Spotify, we are primarily interested in 4 columns; *top year* (int), *genre* (string), *speechiness* (float), and *popularity* (float). *Top year* gives us the year in which the song was a top hit. *Genre* describes the genre of the song, ie: pop, rock, country, etc. *Speechiness* detects the presence of words within a song track, ie: if the song is closer to that of a podcast in which it is entirely speaking and very little to no music it will have a speechiness of 50, 0 if it is entirely music. *Popularity* describes the popularity of a track with 100 being the most popular, 0 being the least. *Top year* allows us to combine the Spotify dataset with the election dataset. We are interested in *genre* because we want to see if there is a correlation between election results and song genres. *Popularity* is a way for us to give weight to each separate song track as the more popular a song is, the more weight it will be given since more people have listened to it and thus the more influence it could’ve had on the election. *Speechiness* is another metric for us to correlate to elections as speechiness is also closely tied to genre and we are curious to see if it has any meaningful predictive power on election results.

For the dataset that contains the election data we are interested in the columns; *year* (int), *state* (string), *party_simplified* (string), *totalvotes* (int), *candidatevotes* (int). The *year* column gives us the year of the election, *state* gives us the corresponding state, *party_simplified* gives us the corresponding party, *totalvotes* is the total number of votes, and *candidatevotes* is the total number of candidate votes. We are interested in the *year* of the election because that is how we will relate our Spotify dataset with the election data. The *state* column will be used to clean our dataset as we are only interested in swing states and thus will be only including states that are widely known to be swing states. These states include Wisconsin, Pennsylvania, New Hampshire, Minnesota, Arizona, Georgia, Virginia, Florida, Michigan, Nevada, Colorado, North Carolina, and Maine.<a name="cite_ref-1"></a>[<sup>1</sup>](#cite_note-1) *Party_simplified* will allow us to know whether or not the resulting vote was Republican or Democrat. *Total_votes* and *candidatevotes* will allow us to know the final voting result.

As stated before we will be combining the Spotify dataset with the election dataset on the year, however by specifying which specific election years we want we can then further reduce the amount of data within both the Spotify and election datasets. After combining both data sets and ensuring there are no null data points, we can then finally get ready to start our EDA.

1. <a name="cite_note-1"></a> [^](#cite_ref-1) https://en.wikipedia.org/wiki/Swing_state

## Spotify Dataset

First, we will import `pandas` and read in our first dataset.

In [257]:
## YOUR CODE TO LOAD/CLEAN/TIDY/WRANGLE THE DATA GOES HERE
## FEEL FREE TO ADD MULTIPLE CELLS PER SECTION 

import pandas as pd

In [258]:
# Read in spotify data

spotify = pd.read_csv("Spotify 2010 - 2019 Top 100.csv")
spotify

title            artist  \
0                     STARSTRUKK (feat. Katy Perry)             3OH!3   
1                       My First Kiss (feat. Ke$ha)             3OH!3   
2                                   I Need A Dollar        Aloe Blacc   
3     Airplanes (feat. Hayley Williams of Paramore)             B.o.B   
4                 Nothin' on You (feat. Bruno Mars)             B.o.B   
...                                             ...               ...   
998                     Strike a Pose (feat. Aitch)  Young T & Bugsey   
999       The London (feat. J. Cole & Travis Scott)        Young Thug   
1000                                            NaN               NaN   
1001                                            NaN               NaN   
1002                                            NaN               NaN   

        top genre  year released       added    bpm  nrgy  dnce   dB  live  \
0       dance pop         2009.0  2022‑02‑17  140.0  81.0  61.0 -6.0  23.0   
1       dance pop         2010.0  2022‑02‑17  138.0  89.0  68.0 -4.0  36.0   
2        pop soul         2010.0  2022‑02‑17   95.0  48.0  84.0 -7.0   9.0   
3     atl hip hop         2010.0  2022‑02‑17   93.0  87.0  66.0 -4.0   4.0   
4     atl hip hop         2010.0  2022‑02‑17  104.0  85.0  69.0 -6.0   9.0   
...           ...            ...         ...    ...   ...   ...  ...   ...   
998     afroswing         2019.0  2020‑08‑20  138.0  58.0  53.0 -6.0  10.0   
999   atl hip hop         2019.0  2020‑06‑22   98.0  59.0  80.0 -7.0  13.0   
1000          NaN            NaN         NaN    NaN   NaN   NaN  NaN   NaN   
1001          NaN            NaN         NaN    NaN   NaN   NaN  NaN   NaN   
1002          NaN            NaN         NaN    NaN   NaN   NaN  NaN   NaN   

       val    dur  acous  spch   pop  top year artist type  
0     23.0  203.0    0.0   6.0  70.0    2010.0         Duo  
1     83.0  192.0    1.0   8.0  68.0    2010.0         Duo  
2     96.0  243.0   20.0   3.0  72.0    2010.0        Solo  
3     38.0  180.0   11.0  12.0  80.0    2010.0        Solo  
4     74.0  268.0   39.0   5.0  79.0    2010.0        Solo  
...    ...    ...    ...   ...   ...       ...         ...  
998   59.0  214.0    1.0  10.0  67.0    2019.0         Duo  
999   18.0  200.0    2.0  15.0  75.0    2019.0        Solo  
1000   NaN    NaN    NaN   NaN   NaN       NaN         NaN  
1001   NaN    NaN    NaN   NaN   NaN       NaN         NaN  
1002   NaN    NaN    NaN   NaN   NaN       NaN         NaN  

[1003 rows x 17 columns]

The dataset has 3 empty rows at the bottom so we'll remove those.

In [259]:
# Make a separate DataFrame for tidying
tidy_spotify = spotify

# Remove empty bottom rows
tidy_spotify = tidy_spotify.iloc[:-3]
tidy_spotify


title              artist  \
0                    STARSTRUKK (feat. Katy Perry)               3OH!3   
1                      My First Kiss (feat. Ke$ha)               3OH!3   
2                                  I Need A Dollar          Aloe Blacc   
3    Airplanes (feat. Hayley Williams of Paramore)               B.o.B   
4                Nothin' on You (feat. Bruno Mars)               B.o.B   
..                                             ...                 ...   
995                                     SICKO MODE        Travis Scott   
996                                      EARFQUAKE  Tyler, The Creator   
997                      Boasty (feat. Idris Elba)               Wiley   
998                    Strike a Pose (feat. Aitch)    Young T & Bugsey   
999      The London (feat. J. Cole & Travis Scott)          Young Thug   

       top genre  year released       added    bpm  nrgy  dnce   dB  live  \
0      dance pop         2009.0  2022‑02‑17  140.0  81.0  61.0 -6.0  23.0   
1      dance pop         2010.0  2022‑02‑17  138.0  89.0  68.0 -4.0  36.0   
2       pop soul         2010.0  2022‑02‑17   95.0  48.0  84.0 -7.0   9.0   
3    atl hip hop         2010.0  2022‑02‑17   93.0  87.0  66.0 -4.0   4.0   
4    atl hip hop         2010.0  2022‑02‑17  104.0  85.0  69.0 -6.0   9.0   
..           ...            ...         ...    ...   ...   ...  ...   ...   
995      hip hop         2018.0  2020‑06‑22  155.0  73.0  83.0 -4.0  12.0   
996      hip hop         2019.0  2020‑06‑22   80.0  50.0  55.0 -9.0  80.0   
997        grime         2019.0  2020‑06‑22  103.0  77.0  89.0 -5.0   9.0   
998    afroswing         2019.0  2020‑08‑20  138.0  58.0  53.0 -6.0  10.0   
999  atl hip hop         2019.0  2020‑06‑22   98.0  59.0  80.0 -7.0  13.0   

      val    dur  acous  spch   pop  top year artist type  
0    23.0  203.0    0.0   6.0  70.0    2010.0         Duo  
1    83.0  192.0    1.0   8.0  68.0    2010.0         Duo  
2    96.0  243.0   20.0   3.0  72.0    2010.0        Solo  
3    38.0  180.0   11.0  12.0  80.0    2010.0        Solo  
4    74.0  268.0   39.0   5.0  79.0    2010.0        Solo  
..    ...    ...    ...   ...   ...       ...         ...  
995  45.0  313.0    1.0  22.0  86.0    2019.0        Solo  
996  41.0  190.0   23.0   7.0  85.0    2019.0        Solo  
997  46.0  177.0    1.0   7.0  68.0    2019.0        Solo  
998  59.0  214.0    1.0  10.0  67.0    2019.0         Duo  
999  18.0  200.0    2.0  15.0  75.0    2019.0        Solo  

[1000 rows x 17 columns]

In [260]:
# Check for NaN values
tidy_spotify.isna().any(axis=None)

False

Since the code returns False, we know we have no other missing data in our dataset!

Next, we need to check the type of each column.

In [261]:
tidy_spotify.dtypes 

title             object
artist            object
top genre         object
year released    float64
added             object
bpm              float64
nrgy             float64
dnce             float64
dB               float64
live             float64
val              float64
dur              float64
acous            float64
spch             float64
pop              float64
top year         float64
artist type       object
dtype: object

All of our numeric columns are already floats, except the `added` column. We will not be using this column, so we can ignore it for now. `top year` can be changed to an integer to allow for easy merging later on. Besides that, we do not need to change the type of any other column.

In [262]:
# Change type of "top year" to int
tidy_spotify['top year'] = tidy_spotify['top year'].astype(int)
tidy_spotify['top year']

/var/folders/q3/5_2c3qgd7hs_m8wkyzqfbjgh0000gn/T/ipykernel_82562/3836981581.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  tidy_spotify['top year'] = tidy_spotify['top year'].astype(int)


0      2010
1      2010
2      2010
3      2010
4      2010
       ... 
995    2019
996    2019
997    2019
998    2019
999    2019
Name: top year, Length: 1000, dtype: int64

Furthermore, we only want to consider the songs from 2011 onward since Spotify was not launched in the US until then.

In [263]:
# Filter only data from 2011 onward 
tidy_spotify = tidy_spotify[tidy_spotify['top year'] >= 2011]
tidy_spotify = tidy_spotify.reset_index() # Reset the index
tidy_spotify

index                                      title              artist  \
0      100                        Rolling in the Deep               Adele   
1      101                       Set Fire to the Rain               Adele   
2      102                               Mr. Saxobeat      Alexandra Stan   
3      103                              What the Hell       Avril Lavigne   
4      104                                   Lighters      Bad Meets Evil   
..     ...                                        ...                 ...   
895    995                                 SICKO MODE        Travis Scott   
896    996                                  EARFQUAKE  Tyler, The Creator   
897    997                  Boasty (feat. Idris Elba)               Wiley   
898    998                Strike a Pose (feat. Aitch)    Young T & Bugsey   
899    999  The London (feat. J. Cole & Travis Scott)          Young Thug   

           top genre  year released       added    bpm  nrgy  dnce   dB  live  \
0       british soul         2011.0  2020‑06‑16  105.0  77.0  73.0 -5.0   5.0   
1       british soul         2011.0  2020‑06‑16  108.0  67.0  60.0 -4.0  11.0   
2          dance pop         2011.0  2020‑06‑16  127.0  93.0  72.0 -4.0  14.0   
3       canadian pop         2011.0  2020‑06‑16  150.0  93.0  58.0 -4.0  14.0   
4    detroit hip hop         2011.0  2020‑06‑16   90.0  70.0  68.0 -8.0  12.0   
..               ...            ...         ...    ...   ...   ...  ...   ...   
895          hip hop         2018.0  2020‑06‑22  155.0  73.0  83.0 -4.0  12.0   
896          hip hop         2019.0  2020‑06‑22   80.0  50.0  55.0 -9.0  80.0   
897            grime         2019.0  2020‑06‑22  103.0  77.0  89.0 -5.0   9.0   
898        afroswing         2019.0  2020‑08‑20  138.0  58.0  53.0 -6.0  10.0   
899      atl hip hop         2019.0  2020‑06‑22   98.0  59.0  80.0 -7.0  13.0   

      val    dur  acous  spch   pop  top year artist type  
0    51.0  228.0   14.0   3.0  84.0      2011        Solo  
1    45.0  243.0    0.0   2.0  81.0      2011        Solo  
2    78.0  195.0    3.0   5.0  69.0      2011        Solo  
3    88.0  221.0    0.0   5.0  80.0      2011        Solo  
4    14.0  304.0   35.0  25.0  74.0      2011         Duo  
..    ...    ...    ...   ...   ...       ...         ...  
895  45.0  313.0    1.0  22.0  86.0      2019        Solo  
896  41.0  190.0   23.0   7.0  85.0      2019        Solo  
897  46.0  177.0    1.0   7.0  68.0      2019        Solo  
898  59.0  214.0    1.0  10.0  67.0      2019         Duo  
899  18.0  200.0    2.0  15.0  75.0      2019        Solo  

[900 rows x 18 columns]

Finally, we will only keep the columns relevant to our research project: `top genre`, `spch`, `pop`, and `top year`.

In [264]:
# Filter only specified columns
tidy_spotify = tidy_spotify[['top genre', 'spch', 'pop', 'top year']]
tidy_spotify

top genre  spch   pop  top year
0       british soul   3.0  84.0      2011
1       british soul   2.0  81.0      2011
2          dance pop   5.0  69.0      2011
3       canadian pop   5.0  80.0      2011
4    detroit hip hop  25.0  74.0      2011
..               ...   ...   ...       ...
895          hip hop  22.0  86.0      2019
896          hip hop   7.0  85.0      2019
897            grime   7.0  68.0      2019
898        afroswing  10.0  67.0      2019
899      atl hip hop  15.0  75.0      2019

[900 rows x 4 columns]

Our Spotify data is now clean!!!

## President Dataset

Now, we will read in our President data.

In [265]:
## YOUR CODE TO LOAD/CLEAN/TIDY/WRANGLE THE DATA GOES HERE
## FEEL FREE TO ADD MULTIPLE CELLS PER SECTION 

president = pd.read_csv("1976-2020-president.csv")
president

year    state state_po  state_fips  state_cen  state_ic        office  \
0     1976  ALABAMA       AL           1         63        41  US PRESIDENT   
1     1976  ALABAMA       AL           1         63        41  US PRESIDENT   
2     1976  ALABAMA       AL           1         63        41  US PRESIDENT   
3     1976  ALABAMA       AL           1         63        41  US PRESIDENT   
4     1976  ALABAMA       AL           1         63        41  US PRESIDENT   
...    ...      ...      ...         ...        ...       ...           ...   
4282  2020  WYOMING       WY          56         83        68  US PRESIDENT   
4283  2020  WYOMING       WY          56         83        68  US PRESIDENT   
4284  2020  WYOMING       WY          56         83        68  US PRESIDENT   
4285  2020  WYOMING       WY          56         83        68  US PRESIDENT   
4286  2020  WYOMING       WY          56         83        68  US PRESIDENT   

                    candidate              party_detailed writein  \
0               CARTER, JIMMY                    DEMOCRAT   False   
1                FORD, GERALD                  REPUBLICAN   False   
2              MADDOX, LESTER  AMERICAN INDEPENDENT PARTY   False   
3     BUBAR, BENJAMIN ""BEN""                 PROHIBITION   False   
4                   HALL, GUS         COMMUNIST PARTY USE   False   
...                       ...                         ...     ...   
4282            JORGENSEN, JO                 LIBERTARIAN   False   
4283            PIERCE, BROCK                 INDEPENDENT   False   
4284                      NaN                         NaN    True   
4285                OVERVOTES                         NaN   False   
4286               UNDERVOTES                         NaN   False   

      candidatevotes  totalvotes   version  notes party_simplified  
0             659170     1182850  20210113    NaN         DEMOCRAT  
1             504070     1182850  20210113    NaN       REPUBLICAN  
2               9198     1182850  20210113    NaN            OTHER  
3               6669     1182850  20210113    NaN            OTHER  
4               1954     1182850  20210113    NaN            OTHER  
...              ...         ...       ...    ...              ...  
4282            5768      278503  20210113    NaN      LIBERTARIAN  
4283            2208      278503  20210113    NaN            OTHER  
4284            1739      278503  20210113    NaN            OTHER  
4285             279      278503  20210113    NaN            OTHER  
4286            1459      278503  20210113    NaN            OTHER  

[4287 rows x 15 columns]

Notice that the "notes" column is NaN in every entry. We need to drop this column. 
After, we can drop rows with at least one NaN value because it is either a candidate that was written in or candidate with no party-- both of which we do not need for our project.

In [266]:
# Make a separate DataFrame for tidying
tidy_president = president

# Drop 'notes' column and NaN rows
tidy_president.drop('notes', axis=1, inplace=True)
tidy_president.dropna(inplace=True)
tidy_president

year      state state_po  state_fips  state_cen  state_ic        office  \
0     1976    ALABAMA       AL           1         63        41  US PRESIDENT   
1     1976    ALABAMA       AL           1         63        41  US PRESIDENT   
2     1976    ALABAMA       AL           1         63        41  US PRESIDENT   
3     1976    ALABAMA       AL           1         63        41  US PRESIDENT   
4     1976    ALABAMA       AL           1         63        41  US PRESIDENT   
...    ...        ...      ...         ...        ...       ...           ...   
4278  2020  WISCONSIN       WI          55         35        25  US PRESIDENT   
4280  2020    WYOMING       WY          56         83        68  US PRESIDENT   
4281  2020    WYOMING       WY          56         83        68  US PRESIDENT   
4282  2020    WYOMING       WY          56         83        68  US PRESIDENT   
4283  2020    WYOMING       WY          56         83        68  US PRESIDENT   

                    candidate              party_detailed writein  \
0               CARTER, JIMMY                    DEMOCRAT   False   
1                FORD, GERALD                  REPUBLICAN   False   
2              MADDOX, LESTER  AMERICAN INDEPENDENT PARTY   False   
3     BUBAR, BENJAMIN ""BEN""                 PROHIBITION   False   
4                   HALL, GUS         COMMUNIST PARTY USE   False   
...                       ...                         ...     ...   
4278            CHARLES, MARK                 INDEPENDENT    True   
4280      BIDEN, JOSEPH R. JR                    DEMOCRAT   False   
4281         TRUMP, DONALD J.                  REPUBLICAN   False   
4282            JORGENSEN, JO                 LIBERTARIAN   False   
4283            PIERCE, BROCK                 INDEPENDENT   False   

      candidatevotes  totalvotes   version party_simplified  
0             659170     1182850  20210113         DEMOCRAT  
1             504070     1182850  20210113       REPUBLICAN  
2               9198     1182850  20210113            OTHER  
3               6669     1182850  20210113            OTHER  
4               1954     1182850  20210113            OTHER  
...              ...         ...       ...              ...  
4278              52     3298041  20210113            OTHER  
4280           73491      278503  20210113         DEMOCRAT  
4281          193559      278503  20210113       REPUBLICAN  
4282            5768      278503  20210113      LIBERTARIAN  
4283            2208      278503  20210113            OTHER  

[3754 rows x 14 columns]

In [267]:
# Check for missing data
tidy_president.isna().any(axis=None) 

False

Since the code returns False, we know we have no other missing data in our dataset!

Now we can check the types of all of our columns.

In [268]:
tidy_president.dtypes

year                 int64
state               object
state_po            object
state_fips           int64
state_cen            int64
state_ic             int64
office              object
candidate           object
party_detailed      object
writein             object
candidatevotes       int64
totalvotes           int64
version              int64
party_simplified    object
dtype: object

All of the numeric columns are already ints or floats, and we do not need to change the type of any other column.

We can now start filtering out data that are not relevant to our project-- since we are using Spotify data from 2011 to 2019, we will only use election data from these years as well. Note that this means we will have data strictly from the 2012 and 2016 elections.

In [269]:
# Filter by year
tidy_president = president[(president['year'] >= 2011) & (president['year'] <= 2019)]
tidy_president


year      state state_po  state_fips  state_cen  state_ic        office  \
3079  2012    ALABAMA       AL           1         63        41  US PRESIDENT   
3080  2012    ALABAMA       AL           1         63        41  US PRESIDENT   
3083  2012     ALASKA       AK           2         94        81  US PRESIDENT   
3084  2012     ALASKA       AK           2         94        81  US PRESIDENT   
3085  2012     ALASKA       AK           2         94        81  US PRESIDENT   
...    ...        ...      ...         ...        ...       ...           ...   
3731  2016  WISCONSIN       WI          55         35        25  US PRESIDENT   
3732  2016    WYOMING       WY          56         83        68  US PRESIDENT   
3733  2016    WYOMING       WY          56         83        68  US PRESIDENT   
3734  2016    WYOMING       WY          56         83        68  US PRESIDENT   
3738  2016    WYOMING       WY          56         83        68  US PRESIDENT   

                          candidate        party_detailed writein  \
3079                   ROMNEY, MITT            REPUBLICAN   False   
3080               OBAMA, BARACK H.              DEMOCRAT   False   
3083                   ROMNEY, MITT            REPUBLICAN   False   
3084               OBAMA, BARACK H.              DEMOCRAT   False   
3085                  JOHNSON, GARY           LIBERTARIAN   False   
...                             ...                   ...     ...   
3731  DE LA FUENTE, ROQUE ""ROCKY""  AMERICAN DELTA PARTY   False   
3732               TRUMP, DONALD J.            REPUBLICAN   False   
3733               CLINTON, HILLARY              DEMOCRAT   False   
3734                  JOHNSON, GARY           LIBERTARIAN   False   
3738             CASTLE, DARRELL L.    CONSTITUTION PARTY   False   

      candidatevotes  totalvotes   version party_simplified  
3079         1255925     2074338  20210113       REPUBLICAN  
3080          795696     2074338  20210113         DEMOCRAT  
3083          164676      300495  20210113       REPUBLICAN  
3084          122640      300495  20210113         DEMOCRAT  
3085            7392      300495  20210113      LIBERTARIAN  
...              ...         ...       ...              ...  
3731            1502     2976150  20210113            OTHER  
3732          174419      258788  20210113       REPUBLICAN  
3733           55973      258788  20210113         DEMOCRAT  
3734           13287      258788  20210113      LIBERTARIAN  
3738            2042      258788  20210113            OTHER  

[549 rows x 14 columns]

We also want to compare only the democrat and republican candidates, as they receive the majority of votes during each election.
To be safe, we will strip and make every value uppercase in the "party_simplified" column.

In [270]:
# Filter by party
tidy_president['party_simplified'] = tidy_president['party_simplified'].str.strip().str.upper()
tidy_president = tidy_president[tidy_president['party_simplified'].isin(['DEMOCRAT','REPUBLICAN'])]
tidy_president

/var/folders/q3/5_2c3qgd7hs_m8wkyzqfbjgh0000gn/T/ipykernel_82562/1523131067.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  tidy_president['party_simplified'] = tidy_president['party_simplified'].str.strip().str.upper()


year          state state_po  state_fips  state_cen  state_ic  \
3079  2012        ALABAMA       AL           1         63        41   
3080  2012        ALABAMA       AL           1         63        41   
3083  2012         ALASKA       AK           2         94        81   
3084  2012         ALASKA       AK           2         94        81   
3088  2012        ARIZONA       AZ           4         86        61   
...    ...            ...      ...         ...        ...       ...   
3719  2016  WEST VIRGINIA       WV          54         55        56   
3723  2016      WISCONSIN       WI          55         35        25   
3724  2016      WISCONSIN       WI          55         35        25   
3732  2016        WYOMING       WY          56         83        68   
3733  2016        WYOMING       WY          56         83        68   

            office         candidate party_detailed writein  candidatevotes  \
3079  US PRESIDENT      ROMNEY, MITT     REPUBLICAN   False         1255925   
3080  US PRESIDENT  OBAMA, BARACK H.       DEMOCRAT   False          795696   
3083  US PRESIDENT      ROMNEY, MITT     REPUBLICAN   False          164676   
3084  US PRESIDENT  OBAMA, BARACK H.       DEMOCRAT   False          122640   
3088  US PRESIDENT      ROMNEY, MITT     REPUBLICAN   False         1233654   
...            ...               ...            ...     ...             ...   
3719  US PRESIDENT  CLINTON, HILLARY       DEMOCRAT   False          188794   
3723  US PRESIDENT  TRUMP, DONALD J.     REPUBLICAN   False         1405284   
3724  US PRESIDENT  CLINTON, HILLARY       DEMOCRAT   False         1382536   
3732  US PRESIDENT  TRUMP, DONALD J.     REPUBLICAN   False          174419   
3733  US PRESIDENT  CLINTON, HILLARY       DEMOCRAT   False           55973   

      totalvotes   version party_simplified  
3079     2074338  20210113       REPUBLICAN  
3080     2074338  20210113         DEMOCRAT  
3083      300495  20210113       REPUBLICAN  
3084      300495  20210113         DEMOCRAT  
3088     2299254  20210113       REPUBLICAN  
...          ...       ...              ...  
3719      713051  20210113         DEMOCRAT  
3723     2976150  20210113       REPUBLICAN  
3724     2976150  20210113         DEMOCRAT  
3732      258788  20210113       REPUBLICAN  
3733      258788  20210113         DEMOCRAT  

[206 rows x 14 columns]

We are only focusing on swing states' data, so we will clean our data to only include those.
From our description above, the 13 most widely known swing states are Wisconsin, Pennsylvania, New Hampshire, 
Minnesota, Arizona, Georgia, Virginia, Florida, Michigan, Nevada, Colorado, North Carolina, and Maine.
To be safe, we will strip and make every value uppercase in the "state" column.

In [271]:
# Filter by state
tidy_president['state'] = tidy_president['state'].str.strip().str.upper()
tidy_president = tidy_president[tidy_president['state'].isin(['WISCONSIN','PENNSYLVANIA','NEW HAMPSHIRE','MINNESOTA',\
                                                              'ARIZONA','GEORGIA','VIRGINA', 'FLORIDA','MICHIGAN','NEVADA',\
                                                              'COLORADO','NORTH CAROLINA', 'MAINE'])]
tidy_president

/var/folders/q3/5_2c3qgd7hs_m8wkyzqfbjgh0000gn/T/ipykernel_82562/2608000910.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  tidy_president['state'] = tidy_president['state'].str.strip().str.upper()


year           state state_po  state_fips  state_cen  state_ic  \
3088  2012         ARIZONA       AZ           4         86        61   
3089  2012         ARIZONA       AZ           4         86        61   
3105  2012        COLORADO       CO           8         84        62   
3106  2012        COLORADO       CO           8         84        62   
3135  2012         FLORIDA       FL          12         59        43   
3136  2012         FLORIDA       FL          12         59        43   
3148  2012         GEORGIA       GA          13         58        44   
3149  2012         GEORGIA       GA          13         58        44   
3202  2012           MAINE       ME          23         11         2   
3203  2012           MAINE       ME          23         11         2   
3219  2012        MICHIGAN       MI          26         34        23   
3220  2012        MICHIGAN       MI          26         34        23   
3225  2012       MINNESOTA       MN          27         41        33   
3226  2012       MINNESOTA       MN          27         41        33   
3255  2012          NEVADA       NV          32         88        65   
3256  2012          NEVADA       NV          32         88        65   
3260  2012   NEW HAMPSHIRE       NH          33         12         4   
3261  2012   NEW HAMPSHIRE       NH          33         12         4   
3292  2012  NORTH CAROLINA       NC          37         56        47   
3293  2012  NORTH CAROLINA       NC          37         56        47   
3319  2012    PENNSYLVANIA       PA          42         23        14   
3320  2012    PENNSYLVANIA       PA          42         23        14   
3383  2012       WISCONSIN       WI          55         35        25   
3384  2012       WISCONSIN       WI          55         35        25   
3407  2016         ARIZONA       AZ           4         86        61   
3408  2016         ARIZONA       AZ           4         86        61   
3431  2016        COLORADO       CO           8         84        62   
3432  2016        COLORADO       CO           8         84        62   
3467  2016         FLORIDA       FL          12         59        43   
3468  2016         FLORIDA       FL          12         59        43   
3474  2016         GEORGIA       GA          13         58        44   
3475  2016         GEORGIA       GA          13         58        44   
3529  2016           MAINE       ME          23         11         2   
3530  2016           MAINE       ME          23         11         2   
3551  2016        MICHIGAN       MI          26         34        23   
3552  2016        MICHIGAN       MI          26         34        23   
3558  2016       MINNESOTA       MN          27         41        33   
3559  2016       MINNESOTA       MN          27         41        33   
3591  2016          NEVADA       NV          32         88        65   
3592  2016          NEVADA       NV          32         88        65   
3597  2016   NEW HAMPSHIRE       NH          33         12         4   
3598  2016   NEW HAMPSHIRE       NH          33         12         4   
3633  2016  NORTH CAROLINA       NC          37         56        47   
3634  2016  NORTH CAROLINA       NC          37         56        47   
3658  2016    PENNSYLVANIA       PA          42         23        14   
3659  2016    PENNSYLVANIA       PA          42         23        14   
3723  2016       WISCONSIN       WI          55         35        25   
3724  2016       WISCONSIN       WI          55         35        25   

            office         candidate           party_detailed writein  \
3088  US PRESIDENT      ROMNEY, MITT               REPUBLICAN   False   
3089  US PRESIDENT  OBAMA, BARACK H.                 DEMOCRAT   False   
3105  US PRESIDENT  OBAMA, BARACK H.                 DEMOCRAT   False   
3106  US PRESIDENT      ROMNEY, MITT               REPUBLICAN   False   
3135  US PRESIDENT  OBAMA, BARACK H.                 DEMOCRAT   False   
3136  US PRESIDENT      ROMNEY, MITT   

Lastly, we are going to focus on the "year", "state", "party_simplified", "totalvotes", and "candidatevotes" columns.


In [272]:
# Reset the index
tidy_president = tidy_president.reset_index()

# Filter by specified column name
tidy_president = tidy_president[["year", "state", "party_simplified", "totalvotes", "candidatevotes"]]
tidy_president

year           state party_simplified  totalvotes  candidatevotes
0   2012         ARIZONA       REPUBLICAN     2299254         1233654
1   2012         ARIZONA         DEMOCRAT     2299254         1025232
2   2012        COLORADO         DEMOCRAT     2569516         1323101
3   2012        COLORADO       REPUBLICAN     2569516         1185243
4   2012         FLORIDA         DEMOCRAT     8474179         4237756
5   2012         FLORIDA       REPUBLICAN     8474179         4163447
6   2012         GEORGIA       REPUBLICAN     3897839         2078688
7   2012         GEORGIA         DEMOCRAT     3897839         1773827
8   2012           MAINE         DEMOCRAT      724758          401306
9   2012           MAINE       REPUBLICAN      724758          292276
10  2012        MICHIGAN         DEMOCRAT     4730961         2564569
11  2012        MICHIGAN       REPUBLICAN     4730961         2115256
12  2012       MINNESOTA         DEMOCRAT     2936561         1546167
13  2012       MINNESOTA       REPUBLICAN     2936561         1320225
14  2012          NEVADA         DEMOCRAT     1014918          531373
15  2012          NEVADA       REPUBLICAN     1014918          463567
16  2012   NEW HAMPSHIRE         DEMOCRAT      710972          369561
17  2012   NEW HAMPSHIRE       REPUBLICAN      710972          329918
18  2012  NORTH CAROLINA       REPUBLICAN     4505372         2270395
19  2012  NORTH CAROLINA         DEMOCRAT     4505372         2178391
20  2012    PENNSYLVANIA         DEMOCRAT     5742040         2990274
21  2012    PENNSYLVANIA       REPUBLICAN     5742040         2680434
22  2012       WISCONSIN         DEMOCRAT     3071434         1620985
23  2012       WISCONSIN       REPUBLICAN     3071434         1410966
24  2016         ARIZONA       REPUBLICAN     2573165         1252401
25  2016         ARIZONA         DEMOCRAT     2573165         1161167
26  2016        COLORADO         DEMOCRAT     2780220         1338870
27  2016        COLORADO       REPUBLICAN     2780220         1202484
28  2016         FLORIDA       REPUBLICAN     9420039         4617886
29  2016         FLORIDA         DEMOCRAT     9420039         4504975
30  2016         GEORGIA       REPUBLICAN     4114732         2089104
31  2016         GEORGIA         DEMOCRAT     4114732         1877963
32  2016           MAINE         DEMOCRAT      771892          357735
33  2016           MAINE       REPUBLICAN      771892          335593
34  2016        MICHIGAN       REPUBLICAN     4799284         2279543
35  2016        MICHIGAN         DEMOCRAT     4799284         2268839
36  2016       MINNESOTA         DEMOCRAT     2944782         1367705
37  2016       MINNESOTA       REPUBLICAN     2944782         1322949
38  2016          NEVADA         DEMOCRAT     1125385          539260
39  2016          NEVADA       REPUBLICAN     1125385          512058
40  2016   NEW HAMPSHIRE         DEMOCRAT      744296          348526
41  2016   NEW HAMPSHIRE       REPUBLICAN      744296          345790
42  2016  NORTH CAROLINA       REPUBLICAN     4741564         2362631
43  2016  NORTH CAROLINA         DEMOCRAT     4741564         2189316
44  2016    PENNSYLVANIA       REPUBLICAN     6115402         2970733
45  2016    PENNSYLVANIA         DEMOCRAT     6115402         2926441
46  2016       WISCONSIN       REPUBLICAN     2976150         1405284
47  2016       WISCONSIN         DEMOCRAT     2976150         1382536

## Combined Dataset

Let's relook at our cleaned datasets.

In [273]:
tidy_spotify

top genre  spch   pop  top year
0       british soul   3.0  84.0      2011
1       british soul   2.0  81.0      2011
2          dance pop   5.0  69.0      2011
3       canadian pop   5.0  80.0      2011
4    detroit hip hop  25.0  74.0      2011
..               ...   ...   ...       ...
895          hip hop  22.0  86.0      2019
896          hip hop   7.0  85.0      2019
897            grime   7.0  68.0      2019
898        afroswing  10.0  67.0      2019
899      atl hip hop  15.0  75.0      2019

[900 rows x 4 columns]

In [274]:
tidy_president

year           state party_simplified  totalvotes  candidatevotes
0   2012         ARIZONA       REPUBLICAN     2299254         1233654
1   2012         ARIZONA         DEMOCRAT     2299254         1025232
2   2012        COLORADO         DEMOCRAT     2569516         1323101
3   2012        COLORADO       REPUBLICAN     2569516         1185243
4   2012         FLORIDA         DEMOCRAT     8474179         4237756
5   2012         FLORIDA       REPUBLICAN     8474179         4163447
6   2012         GEORGIA       REPUBLICAN     3897839         2078688
7   2012         GEORGIA         DEMOCRAT     3897839         1773827
8   2012           MAINE         DEMOCRAT      724758          401306
9   2012           MAINE       REPUBLICAN      724758          292276
10  2012        MICHIGAN         DEMOCRAT     4730961         2564569
11  2012        MICHIGAN       REPUBLICAN     4730961         2115256
12  2012       MINNESOTA         DEMOCRAT     2936561         1546167
13  2012       MINNESOTA       REPUBLICAN     2936561         1320225
14  2012          NEVADA         DEMOCRAT     1014918          531373
15  2012          NEVADA       REPUBLICAN     1014918          463567
16  2012   NEW HAMPSHIRE         DEMOCRAT      710972          369561
17  2012   NEW HAMPSHIRE       REPUBLICAN      710972          329918
18  2012  NORTH CAROLINA       REPUBLICAN     4505372         2270395
19  2012  NORTH CAROLINA         DEMOCRAT     4505372         2178391
20  2012    PENNSYLVANIA         DEMOCRAT     5742040         2990274
21  2012    PENNSYLVANIA       REPUBLICAN     5742040         2680434
22  2012       WISCONSIN         DEMOCRAT     3071434         1620985
23  2012       WISCONSIN       REPUBLICAN     3071434         1410966
24  2016         ARIZONA       REPUBLICAN     2573165         1252401
25  2016         ARIZONA         DEMOCRAT     2573165         1161167
26  2016        COLORADO         DEMOCRAT     2780220         1338870
27  2016        COLORADO       REPUBLICAN     2780220         1202484
28  2016         FLORIDA       REPUBLICAN     9420039         4617886
29  2016         FLORIDA         DEMOCRAT     9420039         4504975
30  2016         GEORGIA       REPUBLICAN     4114732         2089104
31  2016         GEORGIA         DEMOCRAT     4114732         1877963
32  2016           MAINE         DEMOCRAT      771892          357735
33  2016           MAINE       REPUBLICAN      771892          335593
34  2016        MICHIGAN       REPUBLICAN     4799284         2279543
35  2016        MICHIGAN         DEMOCRAT     4799284         2268839
36  2016       MINNESOTA         DEMOCRAT     2944782         1367705
37  2016       MINNESOTA       REPUBLICAN     2944782         1322949
38  2016          NEVADA         DEMOCRAT     1125385          539260
39  2016          NEVADA       REPUBLICAN     1125385          512058
40  2016   NEW HAMPSHIRE         DEMOCRAT      744296          348526
41  2016   NEW HAMPSHIRE       REPUBLICAN      744296          345790
42  2016  NORTH CAROLINA       REPUBLICAN     4741564         2362631
43  2016  NORTH CAROLINA         DEMOCRAT     4741564         2189316
44  2016    PENNSYLVANIA       REPUBLICAN     6115402         2970733
45  2016    PENNSYLVANIA         DEMOCRAT     6115402         2926441
46  2016       WISCONSIN       REPUBLICAN     2976150         1405284
47  2016       WISCONSIN         DEMOCRAT     2976150         1382536

Notice that `tidy_spotify`'s `year` column is from 2011 to 2019. Since we only have election data from 2012 and 2016, we can further tidy filter dataset by year.

In [275]:
tidy_spotify = tidy_spotify[tidy_spotify['top year'].isin([2012, 2016])]
tidy_spotify

top genre  spch   pop  top year
100             british soul   3.0  80.0      2012
101             british soul   4.0  72.0      2012
102                dance pop   8.0  53.0      2012
103  modern alternative rock   4.0  67.0      2012
104               indie rock   3.0  77.0      2012
..                       ...   ...   ...       ...
595                  art pop  11.0  69.0      2016
596  modern alternative rock   9.0  80.0      2016
597  modern alternative rock   6.0  78.0      2016
598                dance pop   5.0  78.0      2016
599                dance pop   5.0  82.0      2016

[200 rows x 4 columns]

We can now merge our newly cleaned data!

In [276]:
tidy = tidy_spotify.merge(tidy_president, left_on='top year', right_on='year')
tidy

top genre  spch   pop  top year  year           state  \
0     british soul   3.0  80.0      2012  2012         ARIZONA   
1     british soul   3.0  80.0      2012  2012         ARIZONA   
2     british soul   3.0  80.0      2012  2012        COLORADO   
3     british soul   3.0  80.0      2012  2012        COLORADO   
4     british soul   3.0  80.0      2012  2012         FLORIDA   
...            ...   ...   ...       ...   ...             ...   
4795     dance pop   5.0  82.0      2016  2016  NORTH CAROLINA   
4796     dance pop   5.0  82.0      2016  2016    PENNSYLVANIA   
4797     dance pop   5.0  82.0      2016  2016    PENNSYLVANIA   
4798     dance pop   5.0  82.0      2016  2016       WISCONSIN   
4799     dance pop   5.0  82.0      2016  2016       WISCONSIN   

     party_simplified  totalvotes  candidatevotes  
0          REPUBLICAN     2299254         1233654  
1            DEMOCRAT     2299254         1025232  
2            DEMOCRAT     2569516         1323101  
3          REPUBLICAN     2569516         1185243  
4            DEMOCRAT     8474179         4237756  
...               ...         ...             ...  
4795         DEMOCRAT     4741564         2189316  
4796       REPUBLICAN     6115402         2970733  
4797         DEMOCRAT     6115402         2926441  
4798       REPUBLICAN     2976150         1405284  
4799         DEMOCRAT     2976150         1382536  

[4800 rows x 9 columns]

We can remove the `top year` column since it is the same as `year`.

In [277]:
tidy.drop(columns=['top year'], inplace=True)
tidy

top genre  spch   pop  year           state party_simplified  \
0     british soul   3.0  80.0  2012         ARIZONA       REPUBLICAN   
1     british soul   3.0  80.0  2012         ARIZONA         DEMOCRAT   
2     british soul   3.0  80.0  2012        COLORADO         DEMOCRAT   
3     british soul   3.0  80.0  2012        COLORADO       REPUBLICAN   
4     british soul   3.0  80.0  2012         FLORIDA         DEMOCRAT   
...            ...   ...   ...   ...             ...              ...   
4795     dance pop   5.0  82.0  2016  NORTH CAROLINA         DEMOCRAT   
4796     dance pop   5.0  82.0  2016    PENNSYLVANIA       REPUBLICAN   
4797     dance pop   5.0  82.0  2016    PENNSYLVANIA         DEMOCRAT   
4798     dance pop   5.0  82.0  2016       WISCONSIN       REPUBLICAN   
4799     dance pop   5.0  82.0  2016       WISCONSIN         DEMOCRAT   

      totalvotes  candidatevotes  
0        2299254         1233654  
1        2299254         1025232  
2        2569516         1323101  
3        2569516         1185243  
4        8474179         4237756  
...          ...             ...  
4795     4741564         2189316  
4796     6115402         2970733  
4797     6115402         2926441  
4798     2976150         1405284  
4799     2976150         1382536  

[4800 rows x 8 columns]

In [298]:
tidy.groupby(['year','state','top genre']).count()

spch  pop  party_simplified  \
year state     top genre                                              
2012 ARIZONA   alternative dance           2    2                 2   
               alternative metal           2    2                 2   
               alternative r&b             4    4                 4   
               art pop                    12   12                12   
               atl hip hop                 4    4                 4   
...                                      ...  ...               ...   
2016 WISCONSIN modern alternative rock     8    8                 8   
               neo mellow                  4    4                 4   
               neo soul                    2    2                 2   
               permanent wave              4    4                 4   
               pop                         6    6                 6   

                                        totalvotes  candidatevotes  
year state     top genre                                            
2012 ARIZONA   alternative dance                 2               2  
               alternative metal                 2               2  
               alternative r&b                   4               4  
               art pop                          12              12  
               atl hip hop                       4               4  
...                                            ...             ...  
2016 WISCONSIN modern alternative rock           8               8  
               neo mellow                        4               4  
               neo soul                          2               2  
               permanent wave                    4               4  
               pop                               6               6  

[768 rows x 5 columns]

In [278]:
# Check for missing values
tidy.isna().any(axis=None)

False

Again, `False` indicates there are no NaN values in the data. Our dataset is ready for analysis!

# Results

## Exploratory Data Analysis

Carry out whatever EDA you need to for your project.  Because every project will be different we can't really give you much of a template at this point. But please make sure you describe the what and why in text here as well as providing interpretation of results and context.

### Section 1 of EDA - please give it a better title than this

Some more words and stuff.  Remember notebooks work best if you interleave the code that generates a result with properly annotate figures and text that puts these results into context.

In [279]:
## YOUR CODE HERE
## FEEL FREE TO ADD MULTIPLE CELLS PER SECTION

tidy.describe()

spch          pop         year    totalvotes  candidatevotes
count  4800.000000  4800.000000  4800.000000  4.800000e+03    4.800000e+03
mean      9.745000    73.970000  2014.000000  3.491030e+06    1.688883e+06
std       8.971998     8.194699     2.000208  2.282838e+06    1.124767e+06
min       3.000000    39.000000  2012.000000  7.109720e+05    2.922760e+05
25%       4.000000    69.000000  2012.000000  2.005787e+06    9.037390e+05
50%       6.000000    74.000000  2014.000000  2.960466e+06    1.393910e+06
75%      10.250000    80.000000  2016.000000  4.733612e+06    2.269228e+06
max      44.000000    91.000000  2016.000000  9.420039e+06    4.617886e+06

### Section 2 of EDA if you need it  - please give it a better title than this

Some more words and stuff.  Remember notebooks work best if you interleave the code that generates a result with properly annotate figures and text that puts these results into context.

In [280]:
## YOUR CODE HERE
## FEEL FREE TO ADD MULTIPLE CELLS PER SECTION

# Ethics & Privacy

- Thoughtful discussion of ethical concerns included
- Ethical concerns consider the whole data science process (question asked, data collected, data being used, the bias in data, analysis, post-analysis, etc.)
- How your group handled bias/ethical concerns clearly described

Acknowledge and address any ethics & privacy related issues of your question(s), proposed dataset(s), and/or analyses. Use the information provided in lecture to guide your group discussion and thinking. If you need further guidance, check out [Deon's Ethics Checklist](http://deon.drivendata.org/#data-science-ethics-checklist). In particular:

- Are there any biases/privacy/terms of use issues with the data you propsed?
- Are there potential biases in your dataset(s), in terms of who it composes, and how it was collected, that may be problematic in terms of it allowing for equitable analysis? (For example, does your data exclude particular populations, or is it likely to reflect particular human biases in a way that could be a problem?)
- How will you set out to detect these specific biases before, during, and after/when communicating your analysis?
- Are there any other issues related to your topic area, data, and/or analyses that are potentially problematic in terms of data privacy and equitable impact?
- How will you handle issues you identified?

_____

The Kaggle dataset we are using analyzes US election results from 1976-2020, based on U.S. states, districts, and on the national level. We assume that since these election results are public nationwide that there wouldn’t be problems with the privacy/terms of use of this data. However, the data itself can have bias since voter turnout has varied from year-to-year, which doesn’t account for the public opinion of everyone in a certain state. Generally, more voter turnout has resulted in elections that were more high-stakes compared to other elections. As well, there are laws that prevent certain demographics from voting: for instance, non-citizens, including legal residents, convicted felons, and some with a mental disability cannot vote based on different states. 

The Spotify dataset was made using Spotify’s web API, and does not rely on the data of their individual users’ listening activity in a way it breaches their personal privacy. However, Spotify wasn’t launched in the U.S. until 2011, and after the trial period ended in January 2012 users were limited to ten hours of streaming per month. We cannot analyze the relationship of music preference and U.S. political preference based on Spotify alone for years before 2011 because of this reason. To answer our research question with best accuracy, we must analyze Spotify listening trends from 2011 and beyond in relation to U.S. election results to show a more direct correlation.

We notice that our dataset of Spotify’s top hits seems to contain a large amount of songs in the “pop” genre, along with hip-hop second (21% and 14% respectively). This could be related to the culture surrounding the time, including the values, conflicts, and social/political climate. Generally, pop music is known for its easy-to-remember lyrics, catchy tune, and feel-good factor that makes people feel happier. Radio stations, advertisements, and the public sphere tend to play pop music for the reason of it being uplifting and easily memorable compared to other genres of music. Therefore, the emphasis on pop music in society can overshadow other music genres that other communities resonate with more for its lyrics and values. Since we are excluding Spotify listening trends prior to 2011, our data will primarily pertain to the U.S. population instead of other countries where Spotify had already been released.

# Team Expectations 


Read over the [COGS108 Team Policies](https://github.com/COGS108/Projects/blob/master/COGS108_TeamPolicies.md) individually. Then, include your group’s expectations of one another for successful completion of your COGS108 project below. Discuss and agree on what all of your expectations are. Discuss how your team will communicate throughout the quarter and consider how you will communicate respectfully should conflicts arise. By including each member’s name above and by adding their name to the submission, you are indicating that you have read the COGS108 Team Policies, accept your team’s expectations below, and have every intention to fulfill them. These expectations are for your team’s use and benefit — they won’t be graded for their details.

* Show up to scheduled meetings consistently
* Evenly assign tasks during meetings
* Complete assigned tasks before meeting 
* Communicate via Discord beforehand if unable to complete assigned tasks or show up to meeting
* Communicate and respond timely (within 1 day)
* Voice opinions respectfully
* Always ask for help if needed!

# Project Timeline Proposal

Specify your team's specific project timeline. An example timeline has been provided. Changes the dates, times, names, and details to fit your group's plan.

If you think you will need any special resources or training outside what we have covered in COGS 108 to solve your problem, then your proposal should state these clearly. For example, if you have selected a problem that involves implementing multiple neural networks, please state this so we can make sure you know what you’re doing and so we can point you to resources you will need to implement your project. Note that you are not required to use outside methods.




| Meeting Date  | Meeting Time| Completed Before Meeting  | Discuss at Meeting |
|---|---|---|---|
| 10/30  |  12:30 PM | Read & Think about COGS 108 expectations; brainstorm topics/questions; find dataset(s)  | Discuss and decide on final project topic; discuss hypothesis; begin background research; submit project proposal | 
| 11/5  |  3 PM |  Do background research on topic | Discuss who will focus on what (Background, Ethics, EDA, Conclusion, etc.); Review proposal feedback and make changes | 
| 11/12  | 2 PM  | Progress on revised proposal  | Work on checkpoint #1; Find new Spotify dataset; Clean/tidy data   |
| 11/19  | 3 PM  | Progress on assigned tasks | Review checkpoint #1 feedback and make changes; Start EDA   |
| 11/27  | 3 PM  | Improve data cleaning; Have EDA ideas | Work on checkpoint #2; Start Analysis |
| 12/3  | 3 PM  | Complete analysis; Draft results/conclusion/discussion | Discuss/edit full project; Work on video |
| 12/11  | Before 11:59 PM  | NA | Turn in Final Project & Group Project Surveys |